dbms/src/Flash/Mpp/TrackedMppDataPacket.h

```diff
@@ -0,0 +1,192 @@
+                size = 0;
+            }
+        }
+    }
+    MemoryTracker * memory_tracker;
+    size_t size = 0;
+};
+
+struct TrackedMppDataPacket
+{
+    explicit TrackedMppDataPacket(const mpp::MPPDataPacket & data, MemoryTracker * memory_tracker)
+        : mem_tracker_wrapper(estimateAllocatedSize(data), memory_tracker)
+    {
+        packet = data;
+    }
+
+    explicit TrackedMppDataPacket()
+        : mem_tracker_wrapper(0, current_memory_tracker)
+    {}
+
+    explicit TrackedMppDataPacket(MemoryTracker * memory_tracker)
+        : mem_tracker_wrapper(0, memory_tracker)
+    {}
+
+    void addChunk(std::string && value)
+    {
+        mem_tracker_wrapper.alloc(value.size());
+        packet.add_chunks(value);
+    }
+
+    void serializeByResponse(const tipb::SelectResponse & response)
+    {
+        mem_tracker_wrapper.alloc(response.ByteSizeLong());
+        if (!response.SerializeToString(packet.mutable_data()))
+            throw Exception(fmt::format("Fail to serialize response, response size: {}", response.ByteSizeLong()));
+    }
+
+    void read(const std::unique_ptr<::grpc::ClientAsyncReader<::mpp::MPPDataPacket>> & reader, void * callback)
+    {
+        reader->Read(&packet, callback);
+        need_recompute = true;
+        //we shouldn't update tracker now, since it's an async reader!!
+    }
+
+    // we need recompute in some cases we can't update memory counter timely, such as async read
+    void recomputeTrackedMem()
+    {
+        if (need_recompute)
+        {
+            mem_tracker_wrapper.freeAll();
+            mem_tracker_wrapper.alloc(estimateAllocatedSize(packet));
+            need_recompute = false;
+        }
+    }
+
+    bool read(const std::unique_ptr<::grpc::ClientReader<::mpp::MPPDataPacket>> & reader)
+    {
+        bool ret = reader->Read(&packet);
+        mem_tracker_wrapper.freeAll();
+        mem_tracker_wrapper.alloc(estimateAllocatedSize(packet));
+        return ret;
+    }
+
+    void switchMemTracker(MemoryTracker * new_memory_tracker)
+    {
+        mem_tracker_wrapper.switchMemTracker(new_memory_tracker);
+    }
+
+    bool hasError() const
+    {
+        return packet.has_error();
+    }
+
+    const ::mpp::Error & error() const
+    {
+        return packet.error();
+    }
+
+    mpp::MPPDataPacket & getPacket()
+    {
+        return packet;
+    }
+
+    MemTrackerWrapper mem_tracker_wrapper;
+    mpp::MPPDataPacket packet;
+    bool need_recompute = false;
+};
+
+struct TrackedSelectResp
+{
+    explicit TrackedSelectResp(tipb::SelectResponse * response)
+        : memory_tracker(response->ByteSizeLong(), current_memory_tracker)
+        , response(response)
+    {}
+
+    void addChunk(std::string && value)
+    {
+        memory_tracker.alloc(value.size());
+        auto * dag_chunk = response->add_chunks();
+        dag_chunk->set_rows_data(value);
+    }
+
+    MemTrackerWrapper memory_tracker;
+    tipb::SelectResponse * response;
```

Comment:
Why save a pointer here? I think `TrackedSelectResp` should be almost the same as `TrackedMppDataPacket`

Reply:
done